ddtrace/internal/coverage/instrumentation_py3_10.py

```diff
@@ -0,0 +1,314 @@
+    process_branches = True
+    exts: t.List[t.Tuple[Instruction, int]] = []
+    while process_branches:
+        process_branches = False
+        for branch in branches:
+            jump_instr = branch.start
+            new_arg = branch.arg
+            jump_instr.arg = new_arg & 0xFF
+            new_arg >>= 8
+            c = 0
+            index = jump_instr.offset >> 1
+
+            # Update the argument of the branching instruction, adding
+            # EXTENDED_ARGs if needed
+            while new_arg:
+                if index and instructions[index - 1].opcode is EXTENDED_ARG:
+                    index -= 1
+                    instructions[index].arg = new_arg & 0xFF
+                else:
+                    ext_instr = Instruction(index << 1, EXTENDED_ARG, new_arg & 0xFF)
+                    instructions.insert(index, ext_instr)
+                    c += 1
+                    # If the jump instruction was a target of another jump,
+                    # make the latest EXTENDED_ARG instruction the target
+                    # of that jump.
+                    if jump_instr.targets:
+                        for target in jump_instr.targets:
+                            if target.end is not jump_instr:
+                                raise (ValueError("Jump instruction is not the end of the branch"))
+                            target.end = ext_instr
+                        ext_instr.targets.extend(jump_instr.targets)
+                        jump_instr.targets.clear()
+                new_arg >>= 8
+
+            # Check if we added any EXTENDED_ARGs because we would have to
+            # reprocess the branches.
+            # TODO[perf]: only reprocess the branches that are affected.
+            # However, this branch is not expected to be taken often.
+            if c:
+                exts.append((ext_instr, c))
+                # Update the instruction offset from the point of insertion
+                # of the EXTENDED_ARGs
+                for instr_index, instr in enumerate(instructions[index + 1 :], index + 1):
+                    instr.offset = instr_index << 1
+
+                process_branches = True
+
+    # Create the new code object
+    new_code = bytearray()
+    for instr in instructions:
+        new_code.append(instr.opcode)
+        new_code.append(instr.arg)
+
+    # Instrument nested code objects recursively
+    for original_offset, nested_code in enumerate(code.co_consts):
+        if isinstance(nested_code, CodeType):
+            new_consts[original_offset], nested_lines = instrument_all_lines(nested_code, trap_func, trap_arg)
+            seen_lines.update(nested_lines)
+
+    new_linetable = update_location_data(code, traps, [(instr.offset, s) for instr, s in exts])
+
+    return (
+        code.replace(
+            co_code=bytes(new_code),
+            co_consts=tuple(new_consts),
+            co_stacksize=code.co_stacksize + 4,  # TODO: Compute the value!
```

Comment:
Same question here. Where did the 4 come from? 😅 Is this just a safe buffer?

Reply:
So the short answer is "to be safe". The long answer is "I don't actually know, @P403n1x87 did that and I'm trying to reverse-engineer my way to the answer and can't." 😓

Reply:
OK, no need to block on this. Just one of the magic numbers I couldn't reverse engineer either. Would be curious to know where it came from if someone knows though.

Reply:
The magic number 4 was chosen for the following reasons:
- when we add our instrumentation, we push additional values to the frame stack (the callable hook object plus its argument); because in theory we might be adding the instrumentation at a point where the stack is at peak usage, it is just safe to add 2 to it (this is an upper bound, but not necessarily the optimal value)
- when I was testing with 2 I was getting segmentation faults that went away by simply increasing the stack size to 4